examples/optimistic-updates-typescript/pages/index.tsx

```diff
@@ -0,0 +1,120 @@
+import * as React from 'react'
+import axios from 'axios'
+
+import {
+  useQuery,
+  useQueryClient,
+  useMutation,
+  QueryClient,
+  QueryClientProvider,
+} from 'react-query'
+import { ReactQueryDevtools } from 'react-query-devtools'
+
+const client = new QueryClient()
+
+export default function App() {
+  return (
+    <QueryClientProvider client={client}>
+      <Example />
+    </QueryClientProvider>
+  )
+}
+
+type Todo = {
+  id: string
+  text: string
+}
+
+type TodoData = {
+  items: readonly Todo[]
+  ts: number
+}
+
+async function fetchTodos(): Promise<TodoData> {
+  const res = await axios.get('/api/data')
+  return res.data
+}
+
+function Example() {
+  const queryClient = useQueryClient()
+  const [text, setText] = React.useState('')
+  const { isFetching, ...queryInfo } = useQuery('todos', fetchTodos)
+
+  const addTodoMutation = useMutation(
+    (newTodo: string) => axios.post('/api/data', { text: newTodo }),
+    {
+      // Optimistically update the cache value on mutate, but store
+      // the old value and return it so that it's accessible in case of
+      // an error
+      onMutate: async newTodo => {
+        setText('')
+        await queryClient.cancelQueries('todos')
+
+        const previousValue = queryClient.getQueryData<TodoData>('todos')
+
+        if (previousValue) {
+          queryClient.setQueryData<TodoData>('todos', {
+            ...previousValue,
+            items: [
+              ...previousValue.items,
+              { id: Math.random().toString(), text: newTodo },
+            ],
+          })
+        }
+
+        return previousValue
+      },
+      // On failure, roll back to the previous value
+      onError: (err, variables, previousValue) => {
+        queryClient.setQueryData('todos', previousValue)
+      },
```

Comment:
@boschni I couldn't get type inference to work properly here. It seems that `TContext` is not inferred by the return value of `onMutate` - it's always `unknown` for me. I know that we can add generics to the `useMutation` call, but it would have to be all 4 of them 😅 .

It works here coincidentally because `unknown` is valid for `setQueryData` if no generic is specified. What I really wanted to do is return a `rollback` function:

 ```
        return () => queryClient.setQueryData('todos', previousValue)
      },
      // On failure, roll back to the previous value
      onError: (err, variables, rollback) => {
        rollback()
      },
```

but this is then not possible ... :(

Reply:
Yeah it seems like TypeScript is not able to infer the context correctly anymore when the context is actually used in any of the onError/onSuccess/onSettled functions. It does infer it correctly when only specifying onMutate. I'll need to figure out why this happens and how this can be fixed

Reply:
It only infers correctly if you do one of the following things:
- explicitly type the variables argument in `onMutate`.
- omit the variables argument in `onMutate`.
- explicitly type the context argument in `onError/onSuccess/onSettled`.

This is probably related to https://github.com/microsoft/TypeScript/issues/26242. Not sure what we can do to fix it.

Reply:
so interestingly, if I define the type of `variables` only on `onMutate`, but not on the `mutationFn`, the variables are still correctly inferred everywhere, including `mutationFn`.

rollback then has the correct type, but it will be potentially undefined:

![Screenshot 2020-12-11 at 19 46 15](https://user-images.githubusercontent.com/1021430/101942301-99e0d700-3be9-11eb-8f15-9077bdcffb41.png)

do you know where the union with undefined is coming from? Because if we can solve that, I think thats a "good enough" approach because we only have to define variables once (`onMutate`). If not, I would just stick with the current approach because it works and people can always annotate their types manually. what do you think?

Reply:
In case of `onError` the context could be `undefined` if the `onMutate` function errored. At `onSuccess` it should always be available though (will fix the types for that). Think defining the variables on `onMutate` is a good compromise.

Apart from that, maybe better to use actual data as context because a rollback function cannot be serialized and persisted: https://react-query-beta.tanstack.com/guides/optimistic-updates

Reply:
Cool, I’ll update to that syntax later today, thanks

Reply:
done: 855b202f